### ISSUE TITLE
Feedback needed: Try out a preview of integrated console support!

### ISSUE THREAD

**Issue by daviwil** [PowerShell/vscode-powershell#311]
I've been working hard over the last few weeks to bring a fully integrated PowerShell console to VS Code via the PowerShell extension.  The work isn't finished yet but I wanted to share a preview release to get some initial feedback which will help guide the final steps to getting this released in 0.8.0 in late October.

The most interesting thing about this feature is that it uses the standard PowerShell console host to provide the interactive console experience.  This is an improvement over what you get in the ISE because you're writing and testing your code in the standard PowerShell host rather than one developed specifically for the editor.

I've only tested this on Windows 10 with PowerShell v5 and v6 and on Windows 7 with PowerShell v3.  Would certainly appreciate to hear whether it works on Windows 8 with PowerShell v4 as well (plus any other combination of Windows/PowerShell versions).

**NOTE:** This only works on Windows right now because I need to do some extra work to make the console launching operate correctly on OS X and Linux.  I might send out an update of the preview within the next few days with full OS X and Linux support.

**Disclaimer: USE AT YOUR OWN RISK**

I have not followed the usual process of getting binaries signed for release because I just wanted to quickly get some feedback on this new feature without a full release process.  As such, I am required to point out that downloading and using this preview is done at your own risk.  Consult the [project license](https://raw.githubusercontent.com/PowerShell/vscode-powershell/master/LICENSE.txt) for full legalese ;)
### How to install and use it
- Install the latest [Visual Studio Code Insiders](https://code.visualstudio.com/insiders) release.  This package does not work with the stable VS Code release (depends on APIs that haven't been released yet)
- **Download [this ZIP file](https://github.com/PowerShell/vscode-powershell/files/512379/ms-vscode.PowerShell-0.7.9.zip)** and extract it into your `$HOME\.vscode-insiders\extensions` folder.
- Load up a PowerShell project and open a `.ps1` or `.psm1` file.  The PowerShell console should open up shortly after and drop you directly into a PowerShell prompt.
### Things to try
- Use the F8 hotkey to execute code from an editor tab like you would in the ISE.  Does it do what you expect?
- Define a function in the console and then try to get IntelliSense for it in the editor
- Call `Get-Credential` or any other cmdlet that never worked right for you in VS Code.
- Use the `psedit` command to open a file (same as in the ISE)
- Try any of the [`$psEditor` API](https://powershell.github.io/PowerShellEditorServices/guide/extensions.html) methods from within the console
- Change the `powershell.developer.powerShellExePath` setting to point to an installation of PowerShell Core on your Windows machine then restart VS Code.  You should get the same console experience for both full PowerShell and the new open-source release of PowerShell.
- Change the `powershell.useX86Host` setting to true if you want to try using the same experience with the 32-bit PowerShell console in a 64-bit Windows install.  This also requires a restart of VS Code.
### Things that don't work
- Probably a lot.  I would not use this for real development yet, it's just a preview of what's coming next.
- PSReadline does not work.  I need to work with @lzybkr to come up with a way to coordinate my usage of the runspace with PSReadline's.  If you load the PSReadline module into your session, all language features will stop working.
- Script debugging mostly doesn't work.  I need to do some hacking magic to make our existing debugging experience work with the console host's debugging implementation.
- Remote PowerShell sessions should work but IntelliSense and other language features will not be aware of the session.  I'm planning to fix this before the official update.
### Updates to the preview

As you report issues, I'll try to push out a few updates to the preview to resolve any major problems.  I'll update this issue with the latest download links as we go.

**Comment by jvheaton** [comment]
Couple things I have noticed.

1. Console threw an error on opening.  Went back an unblocked the zip file and reloaded and then worked.
2. Tab completion in console seems to duplicate when having issues finding the ex. [cd '.\Program Files (x86)\EMET 5.5'cd '.\Program Files (x86)\EMET 5.5']

**Comment by daviwil** [comment]
Hmmm, weird.  I'm surprised the that the ZIP file would have caused a problem.  Do you have the error handy?  If not that's OK.

Not sure why the tab completion would be messed up there, got a screenshot of that?

Thanks for the help!

**Comment by mattmcnabb** [comment]
The ZIP file wouldn't cause an issue - Windows saw this as a remote file and then when you extract, all the extracted files will be remote as well. You just have to unblock the ZIP before you extract.

**Comment by mattmcnabb** [comment]
- Running selections seems to do what is expected and variables, function definitions, etc. are available in the console. F5 still runs in the debugger, but I suppose that's to be expected.
- PSEdit works fine for me.
- $PSEditor also seems to work in the console, at least at a basic level.
- No issues using Get-Credential. Didn't know this was a problem before.
- Ctrl-C doesn't seem to do anything. Could be because this is an editor action for copying?
- Doesn't load my profile. This might be intentional and I haven't dug in to check on this yet.
- Ctrl-C doesn't seem to do anything. Could be because this is an editor action for copying?

**Comment by gerane** [comment]
@daviwil could you do a powershell-preview or similar extension and put it in the marketplace?

**Comment by gerane** [comment]
I'm interested to see if this fixes run selection for me. It was broken with current console. @tyriar was hoping this integration will get that working. It currently doesn't treat code blocks as a whole, but instead each line individually.

**Comment by daviwil** [comment]
@gerane Yeah, I've considered that recently, might try it at some point.

@mattmcnabb Thanks for the notes!  And thanks for reminding me about profile, I forgot to mention that this should be loading the Microsoft.VSCode_profile.ps1 instead of the standard ConsoleHost profile.  If people end up disliking that behavior I'm happy to change it.

Ctrl+C definitely works for me if I'm running a script and want to cancel its execution.  Does it work for you in that case?  If not then it may be an issue I need to investigate.

**Comment by gerane** [comment]
@mattmcnabb ctrl+c has been broken in the integrated terminal for a while. If you highlight text it will copy, but if you highlight text or use ctrl+c in the prompt portion it throws an error. Do you can't break out of running code. It's one of my biggest gripes.

**Comment by gerane** [comment]
@daviwil do you think your integration fixed the ctrl+c to break or did they just recently fix this in there insiders maybe?

**Comment by daviwil** [comment]
@gerane Might have been fixed this week at some point in Insiders, I hadn't tried it before that.  Has been working well for me!

**Comment by gerane** [comment]
@daviwil I'll try it in the morning. There was an update towards the end of the day that I didn't install.

**Comment by gerane** [comment]
@daviwil I realty do think out would be beneficial to have a preview or beta flag for extensions though. I wonder if this has ever been considered.

**Comment by daviwil** [comment]
Not sure, I was looking for an issue about that on the VS Code repo but didn't see one.  I'll file one tomorrow

**Comment by Tyriar** [comment]
You can add "preview" text as in https://marketplace.visualstudio.com/items?itemName=Tyriar.terminal-tabs by adding `"preview": true,` to your [package.json](https://github.com/Tyriar/vscode-terminal-tabs/blob/3d71fa818ddc350a69fbd70a14ceaeb64eb4d205/package.json#L6). It only seems to show on the web though, not in-product so I created https://github.com/Microsoft/vscode/issues/13276

**Comment by gerane** [comment]
@tyriar And I think I know an extension I'll be trying out in the morning. I left a comment on that issue.

**Comment by daviwil** [comment]
@Jaykul reports that text seems to get sent to the terminal in reverse line order (using F8) after he loads his [Profile module](https://github.com/Jaykul/Profile).  Need to investigate why that's happening.

**Comment by gerane** [comment]
Only after profile? Do you think commands that alter your prompt are impacting it?

**Comment by rkeithhill** [comment]
@Jaykul is heavy into developing his PowerLine prompt module.  :-)

**Comment by 4c74356b41** [comment]
wow dude,  this is absolutely stunning, love my vscode even more now!

**Comment by mgreenegit** [comment]
Installed and loaded without issue.  I did need to unblock the zip file, as described above.

**Comment by Jaykul** [comment]
So, my profile.ps1 has LF endings (not CRLF). I modified the `Profile.psd1` to *not* depend on [PowerLine](/Jaykul/PowerLine), and even carefully did: `ipmo Profile; rmo PowerLine;` on one line to make sure PowerLine wasn't loaded and the prompt wasn't changed. 

Even then, hitting F8 in my profile script results in the selected lines being put into the console in reverse order...

**Comment by daviwil** [comment]
@mgreenegit glad to hear it worked after unblocking the files!

@Jaykul I'll give that a shot when I get into the office and see if I can figure out what the deal is

**Comment by dfinke** [comment]
My .zip was not blocked. Had to unblock the start-editorsevice.ps1 though

**Comment by gerane** [comment]
I didn't get to test it today. Had some other things take priority. I did test ctrl+c default behavior @daviwil . If I have the cursor active in the prompt I get an error saying the terminal isn't the active window, but if I click above the prompt it will properly accept ctrl+c.

@jaykul what behavior do you get in the terminal prior? I ask because I have broken behavior with the run selection key binding, and I've wondered if it could be caused by my custom prompt, or if it's broken across the board.

**Comment by gerane** [comment]
@jaykul was also wondering if it could be a line ending issues too.

**Comment by gerane** [comment]
@daviwil just installed it, noticed a few things right off. 

* In terms of profiles, It also does not load the default profile.ps1 for all hosts. It seems to only load the vscode profile, instead of profile.ps1 and vscode profile. 

* It doesn't import PSReadline by default. This breaks keyboard shortcuts, text highlighting with shift, and ctrl + left/right. 

* If you have the prompt selected, it registers as the terminal is not in focus, so ctrl + c to break will not work unless you click the terminal above the prompt. This was present in the terminal already, but still annoying. You get a message saying "Cannot copy terminal selection when terminal does not have focus"

* Still getting odd issues with run selection. It adds in tons of spaces for me. This appears to be caused by PSReadline, as it works as intended if I remove PSReadline. @jaykul did you have PSReadline loaded? it also seems to screw with my line endings and spacing.



This was what I first saw, I will see what else I can find. If you want to talk in more detail or see more let me know @daviwil

**Comment by gerane** [comment]
Also, @Tyriar some of the weird run as selection issues we were discussing, they are being caused by the built in module PSReadline. Maybe we can look a bit more into this and see if we can figure out the cause.

**Comment by gerane** [comment]
@daviwil actually, odd, it doesn't look like it loaded my VSCode profile either. All of my editor commands weren't loaded.

**Comment by gerane** [comment]
@daviwil it doesn't appear to be loading a profile for all for me.

**Comment by gerane** [comment]
@daviwil I have found more problems, especially with Editor Commands, which seem totally broken. Do you want me to keep listing things here or are you ready for separate issues for this. Some of the Editor Command issues:

* PromptForChoice is now redirected to the console. This gives a poor user experience because you have to have the editor pane selected with a ps1 file to trigger an Editor Command. So when it prompts in the console and you start typing, it types in the Editor and not console. It no long feels like an Editor Command, but more like a normal powershell command since you lose the Editor Command Palette experience.

* It cannot recognize PromptForChoice input, and gives the following error

`The pipeline was not run because a pipeline is already running. Pipelines cannot be run concurrently.`

* ctrl + c to break is still broken in some contexts. I was trying to cancel out of PromptForChoice and you can't in the console after starting an Editor Command. It leaves you in a stuck state. You either have to close the console, which opening a new one does not open an integrated one with extension, or you have to reload the Editor.

**Comment by daviwil** [comment]
@gerane I mentioned in the initial post above that PSReadline isn't supported right now because it messes with my ability to use the console's runspace.  I remove it by default when PSES loads so you shouldn't have it in your session unless you loaded it on purpose.  @sgtoj reported issue with using PSReadline VS Code repo [here](https://github.com/Microsoft/vscode/issues/13251) recently.

Regarding profile loading, I forgot to turn that back on before sending out the preview, sorry about that.  That explains why your editor commands weren't showing up.  The change in behavior for the PromptForChoice is an unfortunate result of the new design: since I don't/can't provide my PSHost implementation anymore, you will only ever get the console host's implementation of that method.  My new plan is to provide a more appropriate API for showing editor-native UI for input and choice prompts.  Sucks to have to use two different APIs for this, but I think the benefits of using the console host outweight the inconvenience.

The Ctrl+C issue seems like a VS Code thing, hopefully we can get it worked out in the next couple of releases.

Thanks for the detailed feedback!

**Comment by Tyriar** [comment]
I think ctrl+c is an issue with pty.js

**Comment by gerane** [comment]
@daviwil I would be happy to see a proper API for that instead of Promptforchoice. It has always been a pain to use and had issues. I always felt like it was good for now, but eventually should be replaced with some form of Api. 

It would be nice to build of off $psEditor. Just something like this would be pretty interesting:

```powershell
$Choice = $psEditor.PromptForChoice('Please Select an Option',$Choices)

$Name = $psEditor.Prompt('Please type new name')
```

This is just sort of comparing to the current, but could really be done anyway. Would be nice to not have to deal with building all of the selection lists and just pass $psEditor simple arrays etc.. 

How were you planning on handling alerts and errors? Was it @dfinke that added that in?

**Comment by daviwil** [comment]
Yeah, Doug submitted a pull request for the Show*Message APIs, I'm going to merge it in for 0.8.

Basic Editor Commands are working for me, maybe yours aren't working because they depend on the prompt stuff?  I'll get PSReadline working in 0.9 hopefully, I'll need to enable a hook in PSReadline so that it will give me access to the runspace when I need to run a command.

**Comment by gerane** [comment]
@daviwil yeah, I only tested ones with prompt

**Comment by Jaykul** [comment]
@daviwil @gerane yes! It appears my paste problem is resolved by (re)unloading PSReadLine after importing that Module profile: it sets a bunch of PSReadLine preferences `if($Host.Name -eq "ConsoleHost") {` which of course has to be done by calling cmdlets from PSReadLine, which results in autoloading it.
😖 
Not sure how to fix that -- I'm very wary of custom hosts that are built on ConsoleHost -- how are scripts supposed to avoid breaking?

**Comment by rkeithhill** [comment]
```
Profile script: Host, what's your name?
VSCode terminal: ConsoleHost
Profile script: OK "ConsoleHost", what's your **real** name?
```
😃

**Comment by Jaykul** [comment]
It will make me sad, but now that I've got PS5 everywhere (where PowerShell itself pre-loads the module), maybe I should stop testing to see which host I'm in before using PSReadline, and instead, just test whether it's currently loaded -- if someone has unloaded it for some reason, obviously I don't want to interfere with their devious machinations 😈

**Comment by bcdady** [comment]
I'm not sure if you care to provide this level/depth of backward compatibility, but because I recently got my hands on a rebuilt Win7 with PS v2, I thought I'd give this a go, before upgrading to more current WMF.

I saw the earlier comments about unblocking the zip, and so Unblocked before extracting, and haven't found any extracted files with blocking/blocked issues. I set my ExecutionPolicy to RemoteSigned in the main consolehost, but checked in the extension instance, and noticed it was already at Unrestricted, so I don't think that's inhibiting functionality. Neither F8 hotkey nor `psedit` worked, so I'm guessing the extenion did not load fully/properly.

Here are the error message that appeared on first launch:

```
Remove-Module : No modules were removed. Verify that the specification of modules to remove is correct and those modules exist in the runspace.
At C:\Users\bdady\.vscode-insiders\extensions\ms-vscode.PowerShell-0.7.9\scripts\Start-EditorServices.ps1:68 char:18                                                                                                                                                                                    
+     Remove-Module <<<<  PSReadline                                                                                                                                                                          
+ CategoryInfo          : ResourceUnavailable: (:) [Remove-Module], InvalidOperationException                                                                                                                                                                                                   
+ FullyQualifiedErrorId : Modules_NoModulesRemoved,Microsoft.PowerShell.Commands.RemoveModuleCommand
Module : The 'C:\Users\bdady\.vscode-insiders\extensions\ms-vscode.PowerShell-0.7.9
\modules\PowerShellEditorServices\PowerShellEditorServices.psd1' module cannot be imported  because its manifest contains one or more members that are not valid. The valid manifest members are ('ModuleToProcess', 'NestedModules', 'GUID', 'Author', 'CompanyName', 'Copyright', 'ModuleVersion', 'Description', 'PowerShellVersion', 'PowerShellHostName', 'PowerShellHostVersion', 'CLRVersion', 'DotNetFrameworkVersion', 'ProcessorArchitecture', 'RequiredModules', 'TypesToProcess', 'FormatsToProcess', 'ScriptsToProcess', 'PrivateData', 'RequiredAssemblies', 'ModuleList', 'FileList', 'FunctionsToExport', 'VariablesToExport', 'AliasesToExport', 'CmdletsToExport'). Remove the members that are not valid ('RootModule'), then try to import the module again.                                                                                                                                                                               
At C:\Users\bdady\.vscode-insiders\extensions\ms-vscode.PowerShell-0.7.9\scripts\Start-EditorServices.ps1:167 char:18
+ Import-Module <<<<  PowerShellEditorServices -Version $parsedVersion -ErrorAction Stop
+ CategoryInfo          : InvalidData: (C:\Users\bdady\...orServices.psd1:String) [Import-Module], InvalidOperationException
+ FullyQualifiedErrorId : Modules_InvalidManifestMember,Microsoft.PowerShell.Commands ImportModuleCommand
```

I tried `.vscode-insiders\extensions\ms-vscode.PowerShell-0.7.9\scripts\start.bat ` directly (from the 'TERMINAL'), but got "The system cannot find the path specified"

I also tried `.vscode-insiders\extensions\ms-vscode.PowerShell-0.7.9\scripts\Start-EditorServices.ps1`
but got:

```
Cannot validate argument on parameter 'EditorServicesVersion'. The argument is null or empty. Supply an argument that is not null or empty and then try the command again.                                                                                                                                        
At line:1 char:88                                                                                                                                                                                             
+ .vscode-insiders\extensions\ms-vscode.PowerShell-0.7.9\scripts\Start-EditorServices.ps1 <<<<                                                                                                                
    + CategoryInfo          : InvalidData: (:) [Start-EditorServices.ps1], ParameterBindingValidationException                                                                                                
    + FullyQualifiedErrorId : ParameterArgumentValidationError,Start-EditorServices.ps1
```     

On a positive note: get-credential worked fine for me,

**Comment by daviwil** [comment]
@Jaykul I'm going to provide another way to find out if the console host is being used in an editor, `$psEditor.EditorDetails.Name`or something like that.  In fact, if you see that `$psEditor` is not `$null` then that's a pretty good indication that you're in an editor.  But yeah, you could also check if PSReadline is loaded and don't call the config commands if it's not.  Sorry for the trouble there!

Ultimately I want to get PSReadline working with PSES so hopefully this won't be an issue in the future.  However, that will bring us back to the terminal pasting issues...  We'll have to see why things don't work so cleanly in the VS Code integrated terminal UI.

**Comment by daviwil** [comment]
@bcdady Hey Bryan, thanks for giving this a shot!  The PowerShell extension unfortunately doesn't support PSv2, only PSv3 and above.  A lot of the errors you're seeing are due to me using things that are not available in PSv2.  Was worth a shot though!

**Comment by GHRoss** [comment]
Ooo, sadly I missed this so I couldn't help in testing. Is it due a release soon?

**Comment by daviwil** [comment]
@GHRoss right now I'm shooting for end of October or early November.  Still got a lot of work to finish to make it work well.

**Comment by bcdady** [comment]
Thanks @daviwil - I've updated now to v3, and the earlier reported errors are gone.
I think these symptoms were previously reported, but I'll just confirm my symptoms: that I've not gotten F8 to work, and although `& PROFILE` loads my expected profile script, this script does not load when the TERMINAL extension is initiated.

**Comment by bcdady** [comment]
Is there a shortcut-key, or a way to map keystrokes to jump focus/cursor from editor to console?
In ISE, I frequently use ctrl+D / ctrl+I to jump back and forth, but in Code I'm having to use the mouse to get back and forth.

**Comment by daviwil** [comment]
Regarding the hotkeys, not yet, but I just filed issue #318 to track that.  Also, profiles aren't being loaded in this preview release of the integrated console, I'm adding that back now and might send out an update to the preview in a few days.

**Comment by bcdady** [comment]
While testing with PowerShell v3, I've encountered some code snippets that hang the integrated terminal/console to the point that ctrl+c doesn't recover control. In this case, I resort to clicking the Trash can icon to 'Kill Terminal'. But then I can't find a way to re-start a PowerShell Terminal/console. Clicking the + will add another cmd.exe Terminal/console instance, but the only way (so far) to recover from the hung PowerShell instance is to restart the VS Code - Insider application.

Let me know if you need any further detail, or would prefer I log this as it's own 'Issue'.

**Comment by daviwil** [comment]
Hmmm, weird, snippets are hanging the console?  That's strange, is there a specific one I can try that would reproduce the behavior?

Since releasing this preview I've implemented a "Restart Interactive Console" command which will make it much easier to restart the session without reloading VS Code.  I'm planning to post a new preview ZIP tomorrow with some nice improvements, I'll let you know when that's out.

**Comment by daviwil** [comment]
OK folks, I need your feedback on a dilemma I'm facing with the integrated console work.  I'm becoming increasingly concerned that leveraging the standard "ConsoleHost" is going to be harder to get right than I thought.  I'm not certain that it's a bust yet but I don't want to sink any more time into it if it isn't truly valuable.

What if PSES had it's own console-based host implementation which borrowed as much of the standard ConsoleHost code as possible while having it's own implementation where it mattered?  I don't know for sure yet how much this would cause PSES' host to deviate from the ConsoleHost's behavior, but since this will be written and used as a console application we should be able to reach close compatibility.  The main benefit is that we'd have more control over the hosting environment to provide a more reliable host across the various PowerShell versions we support.

My main concern with continuing the work to hijack the ConsoleHost is that building our editing experience around reflection-based hacks will be more buggy and error prone in the long term.

What do you think?  Is using the standard ConsoleHost valuable enough to continue down that path even if it means the editing/debugging experience might not be as smooth?

**Comment by mattmcnabb** [comment]
Is the ConsoleHost really that valuable at all? I've often thought of it as a limitation and a roadblock to PowerShell adoption. Just look at the number of folks out there on social media that complain about PowerShell, only when questioned their gripes are actually about the ConsoleHost and it's lack of features (improving a lot, I know). This is what drives many to suggest using the ISE while learning PowerShell to take advantage of copy-paste, highlighting, etc.

To balance your question I would say - "What might we lose by dropping the ConsoleHost?" I don't think anyone would fight you on this if the experience will be better/less buggy. However, if there is a significant loss of functionality then that might be a problem. From my limited knowledge, I don't believe there will be any real loss, though.

**Comment by dfinke** [comment]
I don't have enough insight on the custom host v. ConsoleHost. What I know of your ConsoleHost host approach sounded really cool and ambitious. Reflection-based hacks, always come back to bite. The notion of having PSReadline and more in the vs code console is good. Having a consistent console across the boards is optimal. I don't know if that's doable given the legacy issues. Don't know it it's time to break from the that and a better path can be blazed.

**Comment by rkeithhill** [comment]
I don't mind if VSCode-powershell has its own host for debugging and editing REPL work.  If that host has enough implemented to be my primary VSCode integrated terminal even better.  But that means it must handle PSReadLine and posh-git correctly.  :-)  

But if it can't that's OK too as long as I can still bring up PowerShell in the default VSCode terminal window.  Which I currently can with the VSCode 1.7/VSC-PS 0.7.9.  

I think the only issue then would be a bit of education that "hey, the PS Interactive Console" is a REPL for editing and debugging.  For you daily driver for say Git, gcc, etc use the `powershell.exe` terminal window.  Now if we had terminal window tabs, this would make this better IMO.  :-)

**Comment by daviwil** [comment]
Thanks for the feedback guys!

One thing I probably didn't clarify well enough: this will still be a terminal-based app and the REPL experience you get will be very similar to what you see in the current preview.  Also, PSReadLine would play a big role because I'll modify it slightly to drive the console experience.  It basically reimplements the ConsoleHost's input loop so I should be able to take it with some minor modifications and use it to drive my own input loop.  The end goal would be a terminal experience that feels enough like the ConsoleHost + PSReadLine where most users wouldn't be able to tell the difference.

I think I'll do a quick investigation today to see whether it'll be less work to try out this idea than to continue on with the ConsoleHost hacks.

**Comment by dragonwolf83** [comment]
Do you have the same issues with latest Windows 10 Insider builds? Is it improved?

It would be better for Windows 10 long-term if the issues get resolved in the standard ConsoleHost. That doesn't help with legacy, so you may still have to do a custom ConsoleHost for Windows 7-8 based systems. 

For example, Windows 10 gaining 24-bit color support. Are we saying that work will not apply now in PSES unless you add that to the custom host?

**Comment by daviwil** [comment]
@dragonwolf83 the "console host" I'm referring to is `powershell.exe`'s PSHost implementation called `ConsoleHost`. the Windows console host doesn't factor in to the decision because VS Code is hosting `powershell.exe` in a different terminal implementation.

However, the 24-bit color support of the PowerShell console running in VS Code will definitely be limited by VS Code's terminal implementation.  I'm not sure if 24-bit color is supported right now but my guess is that it's not.

**Comment by lzybkr** [comment]
A slightly modified `ConsoleHost` sounds problematic - you're basically forking the code. It would be far better to fix the one console host for your needs.

Note that "better" integration of PSReadline in the console host is problematic. I went with that approach in some early internal builds of V5, it caused problems that weren't easily solved. I'd have to dig some in the SD history to recall what those issues where, but I recall being unhappy removing the tighter integration with PSReadline.

**Comment by daviwil** [comment]
@lzybkr I need to find a solution that will work with PSv3 and up. Improving the PSv6 ConsoleHost doesn't help there unless we can find a way to make it work across my supported range of PowerShell versions.  Let's discuss it next week sometime when you're back, would certainly appreciate your help to determine the best course of action!

**Comment by hillbear** [comment]
As a followup to the Win7 PSv2 issue bcdady brought up, is there an option to disable the integrated console? Sadly, upgrading to a newer WMF will not be an option and basically here for syntax highlighting.

**Comment by daviwil** [comment]
VS Code has PowerShell syntax highlighting built in so the only real value you get on PSv2 is the snippets this extension provides. 

If you install the current release of the PS extension (0.8.0) then it should give you a warning about using an unsupported PowerShell version. If it doesn't then I'll need to fix that.

**Comment by hillbear** [comment]
Thanks! I forgot about the snippets as well.

It does throw the Remove-Module error listed above, not a version warning.

**Comment by daviwil** [comment]
This behavior has been implemented as part of PR #529, closing this issue in favor of #293.  Release should be tomorrow!
